[ENH] JS Client Refactor Part 1 - Refactoring the Client

AlabasterAxe opened the pull request:
## Description of changes

The changes in this PR are described in detail in the "Phase 1" section of this doc: https://docs.google.com/document/d/1DH9Gl5UQeuDr6vyGgoYUw0TkBxITgXIutZSU-ZYdP_E/edit#heading=h.wnaf3x8gqzzl

The TL;DR is that this PR consists of the following changes:
 - turns Collection into a data object (with the exception of the embedding function) and flattens its methods into the top-level ChromaClient
 - Improves the typing of the methods on the client so that we do a better job of handling single vs multiple cases.
 - validates tenant and database on client startup
 - removes some unnecessary code in the client that was just used to enable certain test assertions

This PR corresponds with this issue: https://github.com/chroma-core/chroma/issues/2333

This is the first PR in a series that comprise the whole client refactor:
 - [Part 2: Examples](https://github.com/AlabasterAxe/chroma/pull/1)
 - [Part 3: Tests](https://github.com/AlabasterAxe/chroma/pull/6)
 - [Part 4: Docs](https://github.com/AlabasterAxe/chroma/pull/3)
 - [Full Stack PR](https://github.com/chroma-core/chroma/pull/2542)

## Test plan

- [x] all unit tests are passing 

## Documentation Changes

 - [x] will need to update docs as this is a breaking change with the previous API

atroyn commented:
Adding @codetheweb for review; will cover this this week!

codetheweb commented on clients/js/examples/node/yarn.lock:
should this example use pnpm too?

codetheweb commented on clients/js/src/types.ts:
```diff
@@ -145,15 +173,40 @@ export type ModifyCollectionParams = {
   nResults?: PositiveInteger;
   where?: Where;
-  queryTexts?: string | string[];
+  query: DocQuery;
```
not necessarily opposed to this, but wondering how much we want to try to keep Python/JS interfaces roughly in sync

@atroyn wdyt

atroyn replied:
I think for this PR, we want to keep them in sync. This might be a bit janky for now but I would prefer that we review the interface as a whole between both clients rather than changing it piecemeal in each.

Can we create a new issue to discuss this?

AlabasterAxe replied:
https://github.com/chroma-core/chroma/issues/2483

AlabasterAxe replied:
@atroyn does that apply to the pluralization logic as well? I've added convenience overrides for singular document operations e.g:

```
const resp = getRecords(collection, {
  id: "blah"
});

console.log(resp); // prints "{id: "blah", embedding: [1, 2, 3], document: "whatever" }"
```

codetheweb commented on clients/js/src/ChromaClient.ts:
```diff
@@ -1,15 +1,49 @@
+  validateTenantDatabase,
+  wrapCollection,
+} from "./utils";
 import {
```
```suggestion
import type {
```

codetheweb commented on clients/js/src/ChromaClient.ts:
```diff
@@ -78,10 +104,12 @@ export class ChromaClient {
       database: database,
     });
 
-    // TODO: Validate tenant and database on client creation
```
interesting, wondering if it's worth wrapping the class with a Proxy so we don't have to manually add it to each public method?

AlabasterAxe replied:
Yeah that is a good callout. I think there's a high likelihood that this will get dropped in the future. One thing that the current approach allows is to put certain validations before the `await initPromise;` e.g. 

```TypeScript
if (typeof params.ids === "string") {
  throw new Error(
    "To add a single record, use the singular field names (id, embedding, metadata, document) instead of the plural field names (ids, embeddings, metadatas, documents)",
  );
}
await this.initPromise;
```

Also, ideally, this would only apply to async functions. Do you know if there is a way to detect that in a proxy and only wrap the async functions? Obviously we only have async functions today but I would worry about accidentally making a synchronous function async.

AlabasterAxe replied:
But yeah not opposed to this. My gut reaction would be to hold off on this, but happy to do this in this PR if you think that's the better trade-off.

codetheweb replied:
I think we can leave as-is for now, easy to adjust later

codetheweb commented on clients/js/src/ChromaClient.ts:
````diff
@@ -337,11 +383,377 @@ export class ChromaClient {
+   * ```
+   */
+  async addRecords(
+    collection: Collection,
````
most (all?) methods that deal with documents/records only need the collection's ID, so maybe we could update the typing to reflect that?

```suggestion
    collection: Pick<Collection, "id">,
```

then people can avoid an API call to fetch full collection details if they really want to

AlabasterAxe replied:
Yeah the one thing that we do regularly need is the embedding function so I think it'll beneficial for us to keep Collection as a kind of opaque object that they have to get from the API. My understanding is that there's work underway for the backend to store some representation of embedding function alongside the collection which I think will make this opaque Collection object make more sense.

@atroyn in case I've misunderstood anything here.

codetheweb replied:
personally, as a JS dev, I would assume any object returned from an idiomatic API client library is trivially serializable/de-serializable (`JSON.[stringify|parse]`)
as long as we can hold that assumption that sounds ok to me

codetheweb replied:
alternatively, if every single method only needs the collection ID, I do feel it would be more idiomatic to have methods accept `async addRecords(collection_id: string...)` instead (and store any necessary collection metadata on the client object)

atroyn replied:
We will be able to hold that assumption with EF persist in the works

codetheweb replied:
@AlabasterAxe what do you think about using the collection ID for these methods?

AlabasterAxe replied:
Okay so if I'm understanding what you proposed above, you're saying that the client would maintain a mapping from collection ids to embedding functions and then will just look that up whenever the user supplies a collection id?

One potential issue that could arise is that a user could attempt to call the `addRecords` method even if the embedding function had never been configured for a particular collection.

So hypothetically we could have the following situation:
```TypeScript
/* Run 1 */
const collection = client.createCollection({name: "test"});
console.log(collection.id) // prints 1234

/* Run 2 */
client.addRecords("1234", {/* record contents */}); // How do we know which embedding function to use 
                                                    // here if the client has not gotten the collection in this
                                                    // run of the script?
```

I'll defer to @atroyn here: Should we continue accepting collection objects in the record methods? or should we move to accepting collection ids in the API?

codetheweb replied:
discussed with @atroyn and we settled on keeping the methods accepting collection objects for now

codetheweb commented on clients/js/src/index.ts:
```diff
@@ -35,11 +34,12 @@ export {
-  AddParams,
-  UpsertParams,
-  UpdateParams,
+  AddRecordsParams as addRecordsParams,
```
```suggestion
  AddRecordsParams as AddRecordsParams,
```

codetheweb commented on clients/js/src/index.ts:
```diff
@@ -35,11 +34,12 @@ export {
-  AddParams,
-  UpsertParams,
-  UpdateParams,
+  AddRecordsParams as addRecordsParams,
+  UpsertRecordsParams as UpsertDocumentsParams,
+  UpdateRecordsParams as UpdateDocumentsParams,
```
I think these type names should probably be the same internally and externally?

codetheweb commented on docs/docs.trychroma.com/pages/guides/index.md:
```diff
@@ -271,19 +271,20 @@ client.delete_collection(name="my_collection") # Delete a collection and all ass
-const collection = await client.getCollection({name: "test"}) # Get a collection object from an existing collection, by name. Will raise an exception of it's not found.
-await client.deleteCollection({name: "my_collection"}) # Delete a collection and all associated embeddings, documents, and metadata. ⚠️ This is destructive and not reversible
+const collection = await client.getCollection({ name: "test" }); // Get a collection object from an existing collection, by name. Will raise an exception of it's not found.
+const collection = await client.getOrCreateCollection({ name: "test" }); // Get a collection object from an existing collection, by name. If it doesn't exist, create it.
```
small nit, example won't work as-is because const collection is re-declared

codetheweb commented on docs/docs.trychroma.com/pages/guides/index.md:
```diff
@@ -271,19 +271,20 @@ client.delete_collection(name="my_collection") # Delete a collection and all ass
-await client.deleteCollection({name: "my_collection"}) # Delete a collection and all associated embeddings, documents, and metadata. ⚠️ This is destructive and not reversible
+const collection = await client.getCollection({ name: "test" }); // Get a collection object from an existing collection, by name. Will raise an exception of it's not found.
+const collection = await client.getOrCreateCollection({ name: "test" }); // Get a collection object from an existing collection, by name. If it doesn't exist, create it.
+await client.deleteCollection({ name: "my_collection" }); // Delete a collection and all associated embeddings, documents, and metadata. ⚠️ This is destructive and not reversible
```
I think this implictly can also accept
```suggestion
await client.deleteCollection(collection); // Delete a collection and all associated embeddings, documents, and metadata. ⚠️ This is destructive and not reversible
```
now?

atroyn commented:
Initial review pass looks good, @codetheweb is reviewing docs currently.

@AlabasterAxe could we please split this up into a stack of PRs for easier in-depth review of each part?

codetheweb reviewed (commented):
docs look pretty good to me 👍

AlabasterAxe commented:
> Initial review pass looks good, @codetheweb is reviewing docs currently.
> 
> @AlabasterAxe could we please split this up into a stack of PRs for easier in-depth review of each part?

@atroyn I don't have access to Chroma's Graphite instance so I did this manually. Let me know if that works for you. Also this PR is in kind of a weird state. I don't have permission to push branches to the chroma repo directly so I started it from my personal chroma fork. That means the subsequent PRs are not considered as being against the chroma repo yet but I think that should be okay because once you all give the go ahead I can merge them in quick succession.

Also, it sounds like it may be good to sync about all of the behavior changes this PR includes because it sounds like I may have diverged a bit too much from the Python client. Let me know if you'd like to sync quickly and then I can make the final changes to this PR so it's ready for the final round of reviews.

AlabasterAxe commented:
@atroyn 

I've recorded my thoughts about the various suggested changes to client behavior in these isues:
https://github.com/chroma-core/chroma/issues/2492
https://github.com/chroma-core/chroma/issues/2493
https://github.com/chroma-core/chroma/issues/2494

As discussed yesterday these are just for posterity/a jumping off point for larger conversations about client behavior as opposed to proposals that I'm advocating we implement in the near term. I still need to revert the behavior changes as discussed from yesterday though. Will ping on this PR when those changes are in.

AlabasterAxe commented:
Okay, I've made the requested changes so these PRs should be ready for another look. Specifically:
 - revert the merging of `queryTexts` and `queryEmbedding` into a single `query` param
 - revert the singular field naming
 - revert the parameter-dependent return type (i.e. if user supplies a singular type, unwrap resulting array)

Also, I introduced a some style regressions along the way so I've just added a final formatting PR at the end of the stack.

AlabasterAxe commented:
@atroyn Okay, I've made the requested changes so these PRs should be ready for another look. Specifically:
 - revert the merging of `queryTexts` and `queryEmbedding` into a single `query` param
 - revert the singular field naming
 - revert the parameter-dependent return type (i.e. if user supplies a singular type, unwrap resulting array)

Also, I introduced a some style regressions along the way so I've just added a final formatting PR at the end of the stack.

codetheweb commented on clients/js/src/embeddings/OpenAIEmbeddingFunction.ts:
```diff
@@ -147,10 +147,11 @@ export class OpenAIEmbeddingFunction implements IEmbeddingFunction {
     } catch (e) {
       throw new Error(
-        "Please install openai as a dependency with, e.g. `yarn add openai`",
+        "Please install openai as a dependency with, e.g. `yarn add openai`"
```
somewhat minor, but would be nice to unify the error messages in this file with a single package manager (probably npm)

AlabasterAxe replied:
I wonder if we even need to put an example command? I feel like people should be able to figure out what we mean when we say `please install the openai npm package` WDYT?

codetheweb replied:
seems useful to specify imo 🤷

AlabasterAxe replied:
Done! (here and elsewhere)

codetheweb commented on clients/js/src/ChromaClient.ts:
```diff
@@ -339,11 +384,336 @@ export class ChromaClient {
+    collection: Collection,
+    params: MultiAddRecordsOperationParams,
+  ): Promise<AddResponse>;
+  async addRecords(
```
does this signature (and others like it below) need to be overloaded since the response shape doesn't change?

codetheweb reviewed (commented):
> validates tenant and database on client startup

let's extract this out into a separate PR, we can merge it before this entire stack is approved since it doesn't change any interfaces

also, a quick PR to export types in `package.json`could be merged before this stack is approved as well

otherwise this first PR is looking good! (outside of failing checks)

AlabasterAxe commented:
@codetheweb validation PR here: https://github.com/chroma-core/chroma/pull/2537

This PR is rebased on top of that PR but I'm still using the chroma repo main as its base for this PR.

AlabasterAxe commented:
@codetheweb

I've propagated all of the PR feedback through the stack and put together the "Full stack" PR here: https://github.com/chroma-core/chroma/pull/2542

All PR checks are passing there so I'm thinking that as long as you approve of each of the PRs in the stack, we should be good to merge the Full Stack PR (even though e.g. this PR is failing checks) WDYT?

AlabasterAxe commented on clients/js/src/ChromaClient.ts:
```diff
@@ -339,11 +383,305 @@ export class ChromaClient {
+    collection: Collection,
+    params: AddRecordsParams,
+  ): Promise<AddResponse> {
+    if (typeof params.ids === "string") {
```
NOTE: This check has been reversed in a subsequent stack. Not changing to avoid further PR churn.

jeffchuber commented:
These changes were covered in another PR
